ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -190,17 +205,40 @@ implAddTxs mpEnv@MempoolEnv{mpEnvStateVar, mpEnvLedgerCfg, mpEnvTracer} txs = do
     return $ [(tx, Just err) | (tx, err) <- rejected] ++
              zip accepted (repeat Nothing)
   where
+    MempoolEnv
+      { mpEnvStateVar
+      , mpEnvLedgerCfg
+      , mpEnvCapacity
+      , mpEnvTracer
+      } = mpEnv
+
     traceBatch mkEv size batch
       | null batch = return ()
       | otherwise  = traceWith mpEnvTracer (mkEv batch size)
 
     -- | We first reset 'vrInvalid' to an empty list such that afterwards it
     -- will only contain the /new/ invalid transactions.
-    validateNew :: ValidationResult blk -> ValidationResult blk
-    validateNew res = repeatedly
-        (extendVRNew mpEnvLedgerCfg)
-        txs
-        res { vrInvalid = [] }
+    --
+    -- TODO @intricate: add doc related to the fact that this can block
+    validateNew :: ValidationResult blk -> STM m (ValidationResult blk)
+    validateNew res = repeatedlyM
+      (\tx r -> do
+        initialMempoolSize <- getMempoolSize mpEnv
+        let MempoolCapacity mempoolCap = mpEnvCapacity
+        -- The size of a mempool should never be greater than its capacity.
+        assert (initialMempoolSize <= mempoolCap) $
+          -- Block the thread until the mempool can hold another transaction.
+          --
+          -- TODO @intricate: In this case, 'tx' hasn't been validated yet by
+          -- 'extendVRNew' so we don't actually know whether it's valid.
+          -- Should we only block in the event that we know 'tx' is valid and
+          -- will be added to the mempool?
+          -- As it stands, this code is not bothering to even validate another
+          -- transaction until we know there's enough space in the mempool.
+          check (initialMempoolSize < mempoolCap)
```

Comment:
This is harder than it looks. I think we'll have to rethink our approach to this STM transaction a bit.

The problem at the moment is we're potentially doing a great deal of work before we get to the `check` where we might block. So all of that work could end up being repeated when the STM tx gets retried later. The biggest potential for wasted repeat work is in the `validateIS` in the first part of the transaction, before we come to add new txs to the mempool.

So what approach might work to try to not block if we've done a lot of work already? We could try taking an approach of letting that work commit, and then come back to blocking when it's a much cheaper test to do.

So how about this, we do the `validateIS` but we check if it advanced (really just changed) the tip. This is a proxy for it having done any work. If we did any work, we commit the STM tx and return all the txs yet to add or reject as a result.

Now we go round in a loop, with a new STM tx. So now if the `validateIS` does no more work because the chain didn't change in the meantime (but if it does, it just goes round again), then we go to validate the new txs one by one. But if we find the next one is going to go over the mempool size then what we do should depend on whether this is the first one, or if we managed to make any progress yet (adding or rejecting any). If it's the first one then we can block. This is great, we've done very little work in this STM tx, so STM retry will be cheap. If it's not the first one, then we should again commit what we've done (added & rejected) and return the remaining txs to the outer loop to go round again.

Sounds plausible? Can we think of anything better or neater?

Reply:
- Made some changes yesterday according to @dcoutts review. Still need to properly doc what's going on in several parts of the code as it's not immediately obvious.
- Working on a test as @mrBliss described (https://github.com/input-output-hk/ouroboros-network/pull/890#discussion_r311505517).

Reply:
- It seems that I've now got a working implementation of what @dcoutts described.
- Implemented a test as @mrBliss described (https://github.com/input-output-hk/ouroboros-network/pull/890#discussion_r311505517).